[1단계 - 지하철 노선도 미션] 심바(임선빈) 미션 제출합니다.

0imbean0 opened the pull request:
안녕하세요.
우아한테크코스 3기 심바입니다.

1단계 미션 제출합니다.

### 데모 사이트 - [지하철 노선도](https://bronze-silver-gold-subway.netlify.app)

## 🌱 Step1 기능 구현 목록

### 회원가입

- [x] 유저가 회원가입 버튼을 누른다.
- [x] 유저가 이름을 입력한다.
- [x] 유저가 이메일을 입력한다.
- [x] 유저가 비밀번호, 비밀번호 확인을 입력한다.
- [x] 유저가 확인을 누른다.
- [x] JWT(O) -> 첫 페이지 / JWT(X) -> 로그인 페이지

### 로그인

- [x] 유저가 이메일을 입력한다.
- [x] 유저가 비밀번호를 입력한다.
- [x] 유저가 확인을 입력한다.
- [x] 성공 - 로그인 버튼이 로그아웃 버튼으로 바뀐다.
- [x] 실패 - 빨간 글씨

### 로그아웃

- [x] 유저가 로그아웃 버튼을 누른다.
- [x] 성공 - 로그아웃 버튼이 로그인 버튼으로 바뀐다.

![ezgif com-gif-maker](https://user-images.githubusercontent.com/52202474/112124852-272ec100-8c06-11eb-9d3b-0cb5ddbc67c6.gif)

seojihwan committed:
chore: 환경설정

0imbean0 committed:
chore: 환경 설정 추가

ddongule committed:
docs: Step1 기능 구현 목록 작성

seojihwan committed:
chore: 웹팩 설정 추가

0imbean0 committed:
feat: header 생성 함수 구현

0imbean0 committed:
chore: prettier 옵션 변경

ddongule committed:
refactor: HTML을 Template Literal로 변경

seojihwan committed:
chore: 파일 로더 설정

seojihwan committed:
refactor: login required템플릿 분리

0imbean0 committed:
refactor: template 정리 및 객체 리터럴 선언

ddongule committed:
feat: Route 기능구현

seojihwan committed:
fix: Route 오류 수정

0imbean0 committed:
refactor: route 상수화

0imbean0 committed:
test: 로그인 테스트 작성

ddongule committed:
test: 로그인 테스트 기능 수정

ddongule committed:
refactor: Component 구조 변경

seojihwan committed:
fix: 이름 태그 추가

seojihwan committed:
feat: 회원가입 및 로그인 기능 구현

0imbean0 committed:
refactor: 이벤트 위임 태그 변경

0imbean0 committed:
fix: 로그인 토큰이 받아지지 않는 오류 수정

ddongule committed:
test: 로그아웃 테스트 구현

seojihwan committed:
style: 템플릿 변경

seojihwan committed:
feat: private 라우터 적용

0imbean0 committed:
chore: reload시 해당 페이지 렌더되는 옵션 추가

0imbean0 committed:
feat: accessTokenManager 추가

- 로그인 시 헤더 업데이트
- isLogin hook 추가

ddongule committed:
refactor: html tag 속성 수정

ddongule committed:
refactor: 로그인이 되지 않은 경우, 로그인 페이지로 redirect

ddongule committed:
feat: 로그아웃 기능 구현

seojihwan committed:
refactor: RouteManager 추가

- index.js의 Route가 변경될 때 컴포넌트 렌더하는 로직을 분리

seojihwan committed:
feat: 페이지 이동시마다 토큰 유효성 검증

- 로컬 스토리지 읽기 쓰기기능을 AccessTokenManager가 관리

0imbean0 committed:
feat: 비밀번호 실시간 검증 기능

0imbean0 committed:
fix: access token 검증 조건문 수정

0imbean0 committed:
refactor: setRoute 함수로 변경

ddongule committed:
feat: 이메일 실시간 검증 기능 구현

ddongule committed:
feat: 이름, 이메일, 비밀번호 유효성 검사

seojihwan committed:
refactor: 유효성 검증 로직을 try catch로 변경

0imbean0 committed:
feat: 유효하지 않은 값을 알려주는 UI 구현

0imbean0 committed:
fix: history 저장 오류 해결

- 뒤로가기, load 이벤트 (setRoute -> render)
- privateRouter, publicRouter를 RouteManager의 메서드화

ddongule committed:
style: 전체 style 변경

seojihwan committed:
test: accessToken assertion 문제해결

- url 변경
- 오타 수정

seojihwan committed:
docs: 기능 구현 목록 체크

0imbean0 committed:
refactor: 상수화

- request headers
- regular expression
- message
- standard

seojihwan committed:
chore: 빌드 셋팅

seojihwan committed:
refactor: 회원가입 성공시 자동 로그인

- request 기능 함수로 분리

seojihwan committed:
fix: endOfLine 오류

ysm-dev commented on src/css/ui/modules/form.css:
```diff
@@ -1,8 +1,25 @@
+form input,
+form input:-webkit-autofill {
+  font-size: 1rem !important;
+  background-color: #fff !important;
+  border: 1px solid rgb(0 0 0 / 10%) !important;
+  box-shadow: 4px 4px 10px 0px rgb(226 226 226), 0 0 0 35px white inset !important;
+}
+
+form input:hover,
+form input:focus,
+form input:-webkit-autofill:hover,
+form input:-webkit-autofill:focus {
+  font-size: 1rem !important;
+  background-color: #fff !important;
+  border: 1px solid rgb(78 207 224 / 70%) !important;
+  box-shadow: 4px 4px 10px 0px rgb(78 207 224 / 32%), 0 0 0 35px white inset !important;
```
여기에 !important가 필요한 이유는 무엇인가요?
!important는 특수한 상황이 아니라면 사용을 지양하는것이 좋기 때문에
반드시 사용해야만 하는 이유가있다면 코드리뷰에 참고할 수 있도록 코멘트를 달아주시면 좋을것같아요! 👍

0imbean0 replied:
자동 완성 된 input(email, password)에 대해서 기본 적용된 스타일이 마음에 들지 않아 임의로 CSS를 만들어서 적용했습니다. 
Github 로그인 폼에서도 같은 방법으로 구현되어 있어서 사용해보았습니다. 좋지 않은 문법인가요?

ysm-dev replied:
네 특수한 상황에서만 사용되는 최후의 수단이라고생각하시고,
꼭 필요한 상황이 온다면 주석이나 코드리뷰 코멘트를 달아주시는게 좋을것같아요.

ysm-dev commented on src/js/components/login/index.js:
```diff
@@ -0,0 +1,68 @@
+      }),
+    });
+
+    if (response.status === 500) {
+      throw new ValidationError(ERROR_MESSAGE.LOGIN.EMAIL);
+    }
+
+    if (response.status === 400) {
+      throw new ValidationError(ERROR_MESSAGE.LOGIN.PASSWORD);
+    }
```
실제 statusCode가 이렇게 내려오나요?
일반적으로 HTTP 5XX 에러는 서버 내부에러를 뜻하는데 여기선 그렇지 않은것같네요! 👀

0imbean0 replied:
HTTP 5XX 에러는 서버 내부에러라는 걸 인지하였지만, 현재 API 문서에는 등록되지 않은 이메일에 대한 처리가 되지않아 임의로 처리하였습니다. 그래서 API가 추가되면 수정하려 하였는데, 현재의 사용성을 높이려고 부적절한 처리를 한 것 같습니다. 수정하겠습니다. 😔😢

ysm-dev commented on src/js/components/signup/index.js:
```diff
@@ -0,0 +1,206 @@
+  addEventListeners() {
+    $('#signup-form').addEventListener(
+      'focusout',
+      async ({ target, currentTarget }) => {
+        if (currentTarget['submit'] === target) return;
+
+        if (currentTarget['name'] === target) {
+          const $nameCheck = $('.js-name-check');
+
+          try {
+            this.validateName(target.value);
+            $nameCheck.classList.add('correct');
+            $nameCheck.innerText = CONFIRM_MESSAGE.NAME;
+            this.formValidationFlag.name = true;
+          } catch (error) {
+            if (error instanceof ValidationError) {
+              $nameCheck.classList.remove('correct');
+              $nameCheck.innerText = error.message;
+              this.formValidationFlag.name = false;
+            }
+
+            console.error(error);
+          }
+          return;
+        }
+
+        if (currentTarget['email'] === target) {
+          const $emailCheck = $('.js-email-check');
+
+          try {
+            await this.validateEmail(target.value);
```
이벤트리스너 함수 하나가 너무 큰것같아요! 중복되는 로직이나 작은 단위로 분리할수있을까요?

0imbean0 replied:
관련 코드를 수정 하였습니다. 😀

ysm-dev commented on src/js/stateManagers/AccessTokenManager.js:
```diff
@@ -0,0 +1,33 @@
+    localStorage.removeItem('accessToken');
+  }
+
+  subscribe(listener) {
+    this.listeners.push(listener);
+  }
+
+  notify() {
+    this.listeners.forEach((listener) => listener());
+  }
```
AccessTokenManager가 메뉴 렌더링을 위해 굳이 subscribe, notify 같은 옵저버패턴의 함수들를 가져야할까요?
AccessTokenManager는 뭔가 토큰만 매니징해주면 되는 객체같은데 너무 다양한 일을 한다고 생각되어요. 👀 
(옵저버패턴을 사용하기위해 만든느낌이랄까..?)

0imbean0 replied:
AccessTokenManager에서 Observer Pattern을 사용하는 이유는 App의 하위 컴포넌트에서 App에 있는 renderHeader라는 함수를 사용하기 위해서 입니다. 

Token의 유무는 로그인의 상태와 같은데, 로그인의 상태에 따라 그려주는 Header가 달라집니다.

그래서 Token의 유무(State)를 감지해서 Header를 바꾸어 주는 것이 renderHeader 함수를 모든 컴포넌트에서 호출하는 것보다 코드의 중복을 줄일 수 있어 효율적이라고 생각했습니다.

ysm-dev commented on src/js/utils/fetch.js:
```diff
@@ -0,0 +1,10 @@
+const request = {
+  get: (url, options) => {
+    return fetch(url, { ...options });
+  },
+  post: (url, options) => {
+    return fetch(url, { method: 'POST', ...options });
+  },
+};
```
자주 사용되는 headers나 body의 JSON.stringify 같은 반복되는 것들도 추상화 할 수 있지않을까요?
많이 사용되는 오픈소스 라이브러리인 axios 같은 오픈소스에서는 어떻게 추상화하고있는지 살펴보는것도 도움이 될것같아요 👍

0imbean0 replied:
관련 코드를 수정 하였습니다. 
코드가 더욱 간결해 진 것 같습니다.😀

ysm-dev commented on src/js/components/line/template/main.js:
```diff
@@ -0,0 +1,117 @@
+          />
+        </div>
+        <div class="input-control">
+          <div>
+            <label for="subway-line-color" class="input-label" hidden
+              >간격 시간</label
+            >
+            <input
+              type="text"
+              id="subway-line-color"
+              name="subway-line-color"
+              class="input-field"
+              placeholder="색상을 아래에서 선택해주세요."
+              disabled
+              required
+            />
+          </div>
```
이렇게 div를 한 depth 더 감싸주어야 하는 이유는 무엇인가요? 👀

0imbean0 replied:
이번 단계에서는 아직 모달을 사용하지 않아서 주어진 템플릿 코드를 확인하지 못하였습니다.

다음 단계에서 모달을 사용할 때, 확인하고 수정하겠습니다. 😥

ysm-dev reviewed (changes requested):
1단계 고생하셨습니다! 👏 👏 👏 👏

seojihwan committed:
refactor: state managers 추상화

- route manager의 컴포넌트, 렌더 함수를 App으로 이동
- route manager와 accesstokenmanager를 App의 멤버 변수로 정의
- state manager 추상클래스 정의

0imbean0 committed:
refactor: 회원가입 폼의 focusout 이벤트의 콜백 함수 분리

0imbean0 committed:
refactor: 5XX 에러에 대한 처리 변경

0imbean0 committed:
refactor: fetch params 생성 함수 분리

ysm-dev reviewed (approved):
고생하셨습니다! :)

다음단계에서 봬요! 👏 👏 👏
